include/aws/common/linked_list.h

```diff
@@ -19,82 +19,105 @@
 #include <aws/common/common.h>
 #include <stddef.h>
 
-#define aws_linked_list_of(type, name) struct name {                                                                   \
-    type value;                                                                                                        \
-    struct aws_linked_list_node list;                                                                                  \
-}                                                                                                                      \
-
+/**
+ * from a pointer and a type of the struct containing the node
+ * this will get you back to the pointer of the object. member is the name of
+ * the instance of struct aws_linked_list_node in your struct.
+ */
 #define aws_container_of(ptr, type, member)  \
     ((type *)((uint8_t *)(ptr) - offsetof(type, member)))
 
-#define aws_linked_list_push_back(head, new_node) \
-    aws_linked_list_insert_before((head), (new_node))
-
-#define aws_linked_list_push_front(head, new_node) \
-    aws_linked_list_insert_after((head), (new_node))
-
-#define aws_linked_list_pop_back(head) \
-    aws_linked_list_remove((head)->prev)
-
-#define aws_linked_list_pop_front(head) \
-    head = aws_linked_list_remove(head);
-
-#define aws_linked_list_insert_before(head, new_node) \
-    aws_linked_list_insert_after(head->prev, new_node) 
-
 struct aws_linked_list_node {
     struct aws_linked_list_node *next;
     struct aws_linked_list_node *prev;
 };
 
+struct aws_linked_list {
```

Comment:
This is implemented as a queue, there's no way for a user to run through the nodes unless they're popping them off as they go. I mean, the user can go `node = node->next` but there's no way to know that they've reached the tail node.